packages/mongodb-ts-autocomplete/src/autocompleter.ts

```diff
@@ -0,0 +1,230 @@
+import Autocompleter from '@mongodb-js/ts-autocomplete';
+import type { AutoCompletion } from '@mongodb-js/ts-autocomplete';
+import autocompleteTypes from './fixtures/autocomplete-types';
+
+import type { JSONSchema } from 'mongodb-schema';
+import { toTypescriptTypeDefinition } from 'mongodb-schema';
+
+import {
+  inferCollectionNameFromFunctionCall,
+  compileSourceFile,
+} from './cdt-analyser';
+
+import { CachingAutocompletionContext } from './autocompletion-context';
+import type { AutocompletionContext } from './autocompletion-context';
+
+type MongoDBAutocompleterOptions = {
+  context: AutocompletionContext;
+};
+
+class DatabaseSchema {
+  private collectionSchemas: Record<string, JSONSchema | undefined>;
+
+  constructor() {
+    this.collectionSchemas = Object.create(null);
+  }
+
+  setCollectionNames(collectionNames: string[]): void {
+    // add the missing ones as undefined
+    for (const collectionName of collectionNames) {
+      if (!this.collectionSchemas[collectionName]) {
+        this.collectionSchemas[collectionName] = undefined;
+      }
+    }
+
+    // remove the ones that don't exist anymore
+    const knownCollectionNames = new Set(collectionNames);
+    for (const key of Object.keys(this.collectionSchemas)) {
+      if (!knownCollectionNames.has(key)) {
+        delete this.collectionSchemas[key];
+      }
+    }
+  }
+
+  setCollectionSchema(collectionName: string, schema: JSONSchema): void {
+    this.collectionSchemas[collectionName] = schema;
+  }
+
+  toTypescriptTypeDefinition(): string {
+    const collectionProperties = Object.entries(this.collectionSchemas).map(
+      ([collectionName, schema]) => {
+        const def: string = schema ? toTypescriptTypeDefinition(schema) : `{}`;
+        const lines = def.split(/\n/g).map((line) => `      ${line}`);
+        return `  '${collectionName}': {
```

Comment:
Keep in mind that collection names can contain quotes, newlines, etc.

Should probably be
```suggestion
        return `  ${JSON.stringify(collectionName)}: {
```
?

Reply:
Makes sense. I'll change it throughout.